### PR TITLE
cache: fix stale-while-revalidate and stale-if-error

### LINKED ISSUE #3853: cache: Implement stale-while-revalidate and stale-if-error
I've done a few tests, and those does not seem fully implemented.

https://datatracker.ietf.org/doc/html/rfc5861

### PR DESCRIPTION
**Pull request by flakey5**
Closes #3853

This cleans up the cache interceptor file. This also makes the cache stores store the responses' cache control directives. This was the nicest way I could think of to implement stale-if-error and also sets it up nicely for if we ever allow for more cache control directive extensions.

### CODE REVIEW

**Review by ronag — APPROVED** [review]

File: lib/interceptor/cache.js [hunk 1]
```diff
@@ -1,85 +1,243 @@
...
-  } catch (err) {
-    if (typeof handler.onError === 'function') {
-      handler.onError(err)
+  }
+
+  if (typeof handler.onHeaders === 'function') {
+    // Add the age header
+    // https://www.rfc-editor.org/rfc/rfc9111.html#name-age
+
+    // TODO (fix): What if rawHeaders already contains age header?
```
Comment (metcoder95) [inline lib/interceptor/cache.js:114]: I'd say that we need to keep the request age as source of truth

**Review by metcoder95 — APPROVED** [review]

**Comment by metcoder95** [comment]
types seems to be failing

**Comment by ronag** [comment]
@flakey5 can you rebase?

File: lib/handler/cache-revalidation-handler.js [hunk 2]
```diff
@@ -19,22 +19,30 @@ const DecoratorHandler = require('../handler/decorator-handler')
...
+
   #abort
 
   /**
    * @param {(boolean) => void} callback Function to call if the cached value is valid
    * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   * @param {boolean} allowErrorStatusCodes
    */
-  constructor (callback, handler) {
+  constructor (callback, handler, allowErrorStatusCodes) {
```
Comment (ronag) [inline lib/handler/cache-revalidation-handler.js:45]: Why is this necessary?

Comment (flakey5) [reply lib/handler/cache-revalidation-handler.js:45]: For `stale-if-error`, error status codes when revalidating mean to use the stale response

**Review by metcoder95 — APPROVED** [review]

File: lib/handler/cache-revalidation-handler.js [hunk 3]
```diff
@@ -17,28 +17,45 @@ const assert = require('node:assert')
...
+   */
+  #allowErrorStatusCodes
+
+  #abort
+
+  /**
+   * @param {(boolean) => void} callback Function to call if the cached value is valid
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   * @param {boolean} allowErrorStatusCodes
+>>>>>>> 45c0793e (cache: fix stale-while-revalidate and stale-if-error)
```
Comment (mcollina) [inline lib/handler/cache-revalidation-handler.js:49]: This didn't merge well

File: lib/handler/cache-revalidation-handler.js [hunk 4]
```diff
@@ -59,8 +76,15 @@ class CacheRevalidationHandler {
...
     // https://www.rfc-editor.org/rfc/rfc9111.html#name-handling-a-validation-respo
+<<<<<<< HEAD
     this.#successful = statusCode === 304
     this.#callback(this.#successful, this.#context)
+=======
+    // https://datatracker.ietf.org/doc/html/rfc5861#section-4
+    this.#successful = statusCode === 304 ||
+      (this.#allowErrorStatusCodes && statusCode >= 500 && statusCode <= 504)
+    this.#callback(this.#successful)
+>>>>>>> 45c0793e (cache: fix stale-while-revalidate and stale-if-error)
```
Comment (mcollina) [inline lib/handler/cache-revalidation-handler.js:87]: this didn't merge well

**Review by mcollina — CHANGES_REQUESTED** [review]
There are plenty of merge errors

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**